content/tutorials/aws-persistence.textile

```diff
@@ -0,0 +1,241 @@
+</a>
+
+Update @lambda_function.py@ with this code, remembering to add replace the bucket name with the one you created:
+
+```[python]
+import json, boto3, os, sys, uuid
+from urllib.parse import unquote_plus
+
+s3_client = boto3.client('s3')
+
+def lambda_handler(event, context):
+    
+    event_name = (event['name'])                   # Choosing name field
+    message = (event['data'])                      # Choosing data field
+    data_to_save = event_name + ' : ' + message    # Combining fields
+    print(data_to_save)                            # Checking combination
+    
+    bucket_name = "your-bucket-name-here"          # Target S3 Bucket
+    file_name = "test.txt"                         # Target file
+    lambda_path = "/tmp/" + file_name              # Path of file being download
+    s3_path = "output/" + file_name                # Path of file being uploaded
+    
+    file_object = open(lambda_path, 'a')           # Open downloaded file to append
+    file_object.write("\n" + data_to_save)         # Update downloaded file
+    file_object.close()                            # Save and close edited file
+    
+    s3 = boto3.resource("s3")
+    s3.meta.client.upload_file(lambda_path, bucket_name, file_name)
+```
+
+Once the webhook is set up this code will be invoked every time a message is published. When invoked it will take the data from the JSON Ably sends to Lambda, then open our @test.txt@, append the latest message, save the message and finally upload it back to S3.
+
+Now we're finished in AWS we can set up the Ably half of the workflow.
+
+h2(#installation). Set up Ably app and API Key
+
+We're going to need an API key to use when sending messages to our channel. In your "Ably dashboard":https://ably.com/accounts:
+
+# Hover over @Dashboard@ then select @Apps Dashboard@
+# Click @Create New App@
+# Name your new app
+# Copy the API Key somewhere safe
+
+h2(#webhook). Set up the webhook 
+
+Now we have an app we can set up a webhook to AWS Lambda to trigger when a new message published using the app. In your "Ably dashboard":https://ably.com/accounts:
+
+# Hover over @Dashboard@ then select @Apps Dashboard@
+# Click the name of your app in the URL breadcrumb 
+# Click @Integrations@ then @New Integration Rule@
+# Select @Webhook@ by clicking @Choose@ below it then 
+# Select @AWS Lambda@ by clicking @Choose@ below it
+# Set your webhook settings:
+## Set @AWS Region@ to the the Region your AWS account is located in
+## Enter @ably-persistence@ as the @Function Name@
+## Enter the AWS Key you saved earlier as your @AWS Credentials@
+## Uncheck @Encoding@ (to keep the tutorial as streamlined as possible)
+## Click @Create@
+
+h2(#create-ably-page). Create file to publish messages
+
+To start generating messages to be persisted we are going to set up a basic HTML page that subscribes and publishes to a channel when we click a button.
+
+Create a new @index.html@ file and add:
+
+```[html]
```

Comment:
Using JQuery really dates this example. Also, it might be nice to return a random string rather than the same message each time. E.g.

```javascript
<!DOCTYPE html>
<html lang="en">
  <body>
    <h1>Ably Webhook Lambda Test</h1>
    <button id="publish" type="button" value="Connecting..."></button>
    <div id="result" style="padding: 10px"></div>
  </body>

  <script
    type="text/javascript"
    src="https://cdn.ably.io/lib/ably.min-1.js"
  ></script>
  <!-- Include the latest Ably Library  -->

  <!-- Instance the Ably library  -->
  <script>
    const ably = new Ably.Realtime("ABLY_API_KEY"); /* ADD YOUR API KEY HERE */
    channel =
      ably.channels.get(
        "persistencetest"
      ); /* Channel name to pubish and subscribe to */

    const generateRandomString = (length = 8) =>
      Math.random().toString(20).substr(2, length);

    const btn = document.getElementById("publish");
    const results = document.getElementById("result");

    btn.addEventListener("click", () => {
      channel.publish("message", `${generateRandomString()}`);
    });

    ably.connection.on("connected", () => {
      btn.innerText = "Connected to Ably, click to publish a message";
    });

    channel.subscribe((message) => {
      results.innerHTML += `Message from Ably: ${message.data}`;
    });
  </script>
</html>

```